xbmc/windowing/X11/GLContextEGL.cpp

```diff
@@ -58,154 +60,204 @@ bool CGLContextEGL::Refresh(bool force, int screen, Window glWindow, bool &newCo
-      };
-      m_eglContext = eglCreateContext(m_eglDisplay, m_eglConfig, EGL_NO_CONTEXT, contextAttributes);
-
-      if (m_eglContext == EGL_NO_CONTEXT)
-      {
-        CLog::Log(LOGERROR, "failed to create EGL context\n");
-        return false;
-      }
+      CLog::Log(LOGERROR, "failed to create EGL context\n");
+      Destroy();
+      return false;
+    }
 
-      CLog::Log(LOGWARNING, "Failed to get an OpenGL context supporting core profile 3.2,  \
+    CLog::Log(LOGWARNING, "Failed to get an OpenGL context supporting core profile 3.2,  \
                              using legacy mode with reduced feature set");
-    }
+  }
 
-    if (!eglMakeCurrent(m_eglDisplay, m_eglSurface, m_eglSurface, m_eglContext))
-    {
-      CLog::Log(LOGERROR, "Failed to make context current %p %p %p\n", m_eglDisplay, m_eglSurface, m_eglContext);
-      return false;
-    }
-    XFree(vInfo);
-    retVal = true;
+  if (!eglMakeCurrent(m_eglDisplay, m_eglSurface, m_eglSurface, m_eglContext))
+  {
+    CLog::Log(LOGERROR, "Failed to make context current %p %p %p\n", m_eglDisplay, m_eglSurface, m_eglContext);
+    Destroy();
+    return false;
+  }
+
+  eglGetSyncValuesCHROMIUM = (PFNEGLGETSYNCVALUESCHROMIUMPROC)eglGetProcAddress("eglGetSyncValuesCHROMIUM");
+
+  m_usePB = false;
+  return true;
+}
+
+bool CGLContextEGL::CreatePB()
+{
+  const EGLint configAttribs[] =
+  {
+    EGL_SURFACE_TYPE, EGL_PBUFFER_BIT,
+    EGL_BLUE_SIZE, 8,
+    EGL_GREEN_SIZE, 8,
+    EGL_RED_SIZE, 8,
+    EGL_DEPTH_SIZE, 8,
+    EGL_RENDERABLE_TYPE, EGL_OPENGL_BIT,
+    EGL_NONE
+  };
+
+  const EGLint pbufferAttribs[] =
+  {
+    EGL_WIDTH, 9,
+    EGL_HEIGHT, 9,
+    EGL_NONE,
+  };
+
+  Destroy();
+
+  if (eglGetPlatformDisplayEXT)
+  {
+    m_eglDisplay = eglGetPlatformDisplayEXT(EGL_PLATFORM_X11_EXT,(EGLNativeDisplayType)m_dpy,
+                                            NULL);
   }
   else
+    m_eglDisplay = eglGetDisplay((EGLNativeDisplayType)m_dpy);
+
+  if (m_eglDisplay == EGL_NO_DISPLAY)
   {
-    CLog::Log(LOGERROR, "EGL Error: vInfo is NULL!");
+    CLog::Log(LOGERROR, "failed to get egl display");
+    return false;
+  }
+  if (!eglInitialize(m_eglDisplay, NULL, NULL))
+  {
+    CLog::Log(LOGERROR, "failed to initialize egl\n");
+    Destroy();
+    return false;
+  }
+  if (!eglBindAPI(EGL_OPENGL_API))
+  {
+    CLog::Log(LOGERROR, "failed to initialize egl");
+    Destroy();
+    return false;
   }
 
-  eglGetSyncValuesCHROMIUM = (PFNEGLGETSYNCVALUESCHROMIUMPROC)eglGetProcAddress("eglGetSyncValuesCHROMIUM");
-  return retVal;
+  EGLint numConfigs;
+
+  eglChooseConfig(m_eglDisplay, configAttribs, &m_eglConfig, 1, &numConfigs);
+  m_eglSurface = eglCreatePbufferSurface(m_eglDisplay, m_eglConfig, pbufferAttribs);
+  if (m_eglSurface == EGL_NO_SURFACE)
+  {
+    CLog::Log(LOGERROR, "failed to create EGL window surface %d", eglGetError());
+    Destroy();
+    return false;
+  }
+
+  m_eglContext = eglCreateContext(m_eglDisplay, m_eglConfig, EGL_NO_CONTEXT, NULL);
```

Comment:
Why don't you include the context attributes here?

IMO you should also request 3.2 core profile or 2.0 as fallback like in the non-pbuffer case.

As a (unintentional?) side-effect, this seems to break the fallback to GLX when GL 2.0 isn't supported. I know that moving forward we should require 2.0 (or even 3.2?) anyway, but it would be good to save this for v19 and not break functionality needlessly.